### ISSUE TITLE
Better RegExp syntactic sugar/UX

### ISSUE THREAD

**Issue by richardxia** [sifive/wake#528]
Currently it is very unwieldy to try to combine RegExps and Strings, especially when the Strings may have different runtime values. Consider the following:

```wake
def myFunc outputDirString filesToMatch =
    def pattern = regExpCat (outputDirString.quote, `/(include|lib)/.*`, Nil)
    ...
```

Having sugar for better handling this would be nicer, such as adding an interpolation syntax that can automatically quote strings or making it easier to just concatenate two RegExp objects together. Here's some example syntaxes:

```wake
`{outputDirString}/(include|lib)/.*`
outputDirString.quote + `/(include|lib)/.*`
regExpCat2 outputDirString.quote `/(include|lib)/.*`
```

**Comment by terpstra** [comment]
```
`{outputDirString}/(include|lib)/.*`
```
Interesting. I like this proposal a lot, except for the problem that `{}` is already used in the regular expression language itself. Maybe `${}` ala scala? On the other hand, `x{5,7}` is pretty uncommon in regular expression land. I would rather not have string and regular expression interpolation used different escapes.

```
outputDirString.quote + `/(include|lib)/.*`
```
I guess this is an option. Not super thrilled, though.

```
regExpCat2 outputDirString.quote `/(include|lib)/.*`
```
This seems no better than the current solution, except that it avoids Nil. When/if Nil becomes unnecessary for lists, I think this has no advantage left, so I vote against this option.

**Comment by terpstra** [comment]
On the interpolation side of things, there are two sorts of values you might want to dump into a regular expression:
- another regular expression
- a string which gets quoted

So maybe there need to be two distinct syntaxes for that?
```
def path = "xyz"
def pattern = `.*\.(c|h|cpp|cc|hpp)`
def myRegExp = `foo/${path}/%{pattern}`

**Comment by terpstra** [comment]
OTOH, if we only have ONE capability, it should be to insert another regular expression, since that is more powerful:
```
def path = "xyz"
def pattern = `.*\.(c|h|cpp|cc|hpp)`
def myRegExp = `foo/{path.quote}/{pattern}`
```

**Comment by richardxia** [comment]
I, too, prefer the interpolation idea, but I wanted to throw out a few other ideas.

Re: two different interpolation syntaxes, I feel like I'm not a fan of that, since it's harder to remember which symbol is used for which. I would possibly even prefer not having sugar for automatic quoting, and instead just having a single syntax for interpolation:

```wake
def path = "xyz"
def pattern = `.*\.(c|h|cpp|cc|hpp)`
def myRegExp = `foo/{path | quote}/{pattern}`
```

**Comment by richardxia** [comment]
One other nice thing about the single syntax is that it can kind of be viewed as similar as Python's [f-string formatters](https://www.python.org/dev/peps/pep-0498/#format-specifiers), which look like this (http://cis.bentley.edu/sandbox/wp-content/uploads/Documentation-on-f-strings.pdf):

```python
x = 4.5 print(f'This will print out the variable x: {x:.3f}')
```

It uses the `:` to separate the value being formatted from the formatter options, and the options are type-specific, so custom formatters and formatter options can be defined on custom types. In Wake, either the `.` notation or the `|` notation for calling functions would work here.

**Comment by richardxia** [comment]
Also, re: `{}` and escapes, what about using the same rule for interpolated strings with a custom interpolation sigil?

```wake
`%path/to/%{inputFile}/main\.(c|h){2,5}%`
```

**Comment by terpstra** [comment]
I actually really regret the custom interpolation sigil support I added for multiline strings. It's a real problem for other tools that want to understand wake syntax, because it cannot be supported with a context free grammar. This is a problem for vscode.

Re: automatic escape of strings to regular expressions, I think I agree. Explicitly calling `quote` (maybe `escape` would have been a better name?) does not seem cumbersome and actually improves understanding as to what is happening in that expression.

So that leaves only the issue of: what to do about `{}`s that are used for a real regular expression. Also really awkward is that `\{` is no good, because that is used to get a literal `{` into a regular expression. Unless we descend to `\\{` which is clearly mad.

**Comment by terpstra** [comment]
I'll note that `${xyz}` is actually a pretty safe choice, because `$` in a regular expression can only occur either at the end of the regular expression, or immediately before a `)`.

**Comment by richardxia** [comment]
> I'll note that `${xyz}` is actually a pretty safe choice, because `$` in a regular expression can only occur either at the end of the regular expression, or immediately before a `)`.

Does Google's regular expression library treat `$` as "end of line" or "end of string"? I know that in Ruby, which mostly follows Perl's regex notation, `$` is only treated as "end of line" and not "end of string", so if the input string is a multiline string, then `$\nStart of new line` is a perfectly valid regex. That said, this confuses everyone who uses Ruby.

**Comment by terpstra** [comment]
I've configured RE2 so that `$` always means end of string. If you want end of line, use `\n`.

I just realized that there is a minor technical hurdle, but also major benefit, with the interpolation idea: `({subexpr})*`

Naively, I'd want to turn that into: 
```
regExpCat (`(`, subexpr, `)`, Nil
```

Unfortunately, `(` is clearly not a legal regular expression. What's interesting, though, is that if you replaced all the interpolations with just a `.` and then tested the remaining string (ie: `(.)`) you can be ahead-of-time certain that the regular expression is legal. 

In other words:
- `({subexpr}` should give you a compile-time error.
- interpolation is strictly more powerful than what regExpCat offers today

**Comment by terpstra** [comment]
So I'm leaning towards:
```
def path = "xyz"
def pattern = `.*\.(c|h|cpp|cc|hpp)`
def myRegExp = `foo/${path.quote}/${pattern}`
```

It's unfortunate that regular expression interpolation syntax does not match string interpolation syntax, but I think the simplicity of a fixed escape that does not to conflict with standard regular expression syntax outweighs that. If anything, I'd lean towards changing string interpolation to match regular expression interpolation and scala... ;-)

**Comment by richardxia** [comment]
> So I'm leaning towards:
> 
> ```
> def path = "xyz"
> def pattern = `.*\.(c|h|cpp|cc|hpp)`
> def myRegExp = `foo/${path.quote}/${pattern}`
> ```
> 
> It's unfortunate that regular expression interpolation syntax does not match string interpolation syntax, but I think the simplicity of a fixed escape that does not conflict with standard regular expression syntax outweighs that. If anything, I'd lean towards changing string interpolation to match regular expression interpolation and scala... ;-)

I think I'd agree with everything you said here. I'd also add that the `${}` syntax is used by bash, make, and JavaScript, so it's probably more familiar than just `{}`.

I'd still say that auto-escaping based on the inferred type would be nicer, but I get that the current type system would not necessarily be able to handle that.

**Comment by terpstra** [comment]
> I'd still say that auto-escaping based on the inferred type would be nicer, but I get that the current type system would not necessarily be able to handle that.

I could make that work. I just thought it would be less explicit, were the compiler to automagically behave differently based on the type supplied inside `${}`.

In fact, wake String interpolation originally supported any type, and ran `format` on anything that was not a String. We changed this, because it was deemed too magical and easy to mess up, by accidentally passing the wrong thing.
